### ISSUE TITLE
(aoe): enhance and refactor the buffer related code

### ISSUE THREAD

**Issue by XuPeng-SH** [matrixorigin/matrixone#119]
Right now the managed node of memory is a bare byte slice, and each node has the following known usage scenarios (not all):
- Used as a writable vector internal buffer, which is used for column data insertion
  1. Allocated with a capacity from the memory pool
  2. Construct a vector with the allocated node as the internal buffer
  3. Copy bytes into the underlying buffer
  4. Marshall the buffer to a file writer
  5. Free the memory from the memory pool
- Used as read-only vector internal buffer, which is used for the column data scan
  1. Allocated with a capacity from the memory pool
  2. Load buffer data from a file reader
  3. Unmarshall the buffer data as a vector instance
  4. Free the memory from the memory pool
- Used as read-only index internal buffer, which is used for the index scan
  1. Allocated with a capacity from the memory pool
  2. Load buffer data from a file reader
  3. Unmarshall the buffer data as an index instance
  4. Free the memory from the memory pool

Basically, all scenarios have some commonalities. Currently, the buffer application layer handles operations like serialization and deserialization in their respective. These can all be abstracted out and processed internally to make the application layer more convenient to use.
